### PR TITLE
Avocado Job Diff [V3]

### PR DESCRIPTION
**Pull request by apahim**
v3:
- Highlight section headers with `rst` markers.
- Use `nooption` instead of `-option` for `--diff-filter`.
- Introduce `--create-reports` mode, where we suppress the embedded diff tool and create tmp files with the content to be used in external diff tools. 
- Fix issue when using job results directories to identify jobs.
- Encoding improvements.
- Resources optimizations by not creating intermediary lists with all the content.
- Basic functional tests.

v2: #1369 
- Use the paginator for the diff.
- Allow an optional destination file parameter to be passed with `--html`.
- time formatted as in human-readable plugin.
- Pass a job-results directory as a parameter instead of a SHA1.

v1: #1365 
- Use only sections that have any diff.
- Remove the title from the `legends` in html page.
- Fix the diff-filter argparse.
- `--open-html` to enable `--html`.
- Argparse help messages improved.
- Be explict regarding the time unit (secs).

### CODE REVIEW

**Comment by adereis** [comment]
Working well here. Just a few really minor things:

 * It's leaving behind an empty directory in `/var/tmp/`
 * Empty line at beginning of the files generated when using `--external-tool`
 * When using `--html` and/or `--open-browser`, the stdout report is still generated. Besides being undesired IMO, the report will be open in a browser only after we reach the end of the report inside a paginator.
 * s/Html diff file/HTML diff file/
 * In the help message for `--html`, I don't think `-` should me mentioned. It was copy/pasted from the `run --html` option help, where `-` is used with the other output formats.

With the minor issues above fixed, I would consider this feature ready (as far as usability is concerned -- I haven't reviewed the code).

**Comment by apahim** [comment]
@adereis, I updated this PR. Some comments:

 * It's leaving behind an empty directory in `/var/tmp/`

Could you expand this topic? I don't see the directory left behind.

 * Empty line at beginning of the files generated when using `--external-tool`

All sections have a '\n' before the header to separate from the previous section. We could drop the first '\n' from the results, but I'd prefer to add a header to the files. I added **Avocado Job Report** as the first line.

 * When using `--html` and/or `--open-browser`, the stdout report is still generated. Besides being undesired IMO, the report will be open in a browser only after we reach the end of the report inside a paginator.

I disabled the standard unified diff when `--html` is used.

 * s/Html diff file/HTML diff file/

Done.

 * In the help message for `--html`, I don't think `-` should me mentioned. It was copy/pasted from the `run --html` option help, where `-` is used with the other output formats.

Done.

File: avocado/plugins/diff.py [hunk 1]
```diff
@@ -0,0 +1,427 @@
...
+            with open(config_file, 'r') as f:
+                return f.readlines()
+        except:
+            return ['Not found\n']
+
+    def _get_sysinfo(self, resultsdir, pre_post):
+        sysinfo_dir = os.path.join(resultsdir, 'sysinfo', pre_post)
+        sysinfo = []
+        for path, subdirs, files in os.walk(sysinfo_dir):
+            for name in files:
```
Comment (ldoktor) [inline avocado/plugins/diff.py]: The files are not sorted, making it produce diffs on the same jobs (when the dir is copied). Would you please add `sorted` here?

Comment (ldoktor) [inline avocado/plugins/diff.py]: ...hopefully the paths are...

File: avocado/plugins/diff.py [hunk 2]
```diff
@@ -0,0 +1,427 @@
...
+
+        parser.add_argument('--paginator',
+                            choices=('on', 'off'), default='on',
+                            help='Turn the paginator on/off. '
+                            'Current: %(default)s')
+
+        parser.add_argument('--external-tool', action='store_true',
+                            help='Instead of show the diff using Avocado '
+                            'Diff utility, create temporary files to use '
+                            'in a custom diff tool.')
```
Comment (ldoktor) [inline avocado/plugins/diff.py]: Based on the description I'd call this `store-reports` or something like that. From an option called `--external-tool` I'd expect it to run the tool, eg:

```
avocado diff 123 321 --external-tool meld
Creating diff files /tmp/123 /tmp/321
Running meld /tmp/123 /tmp/321
```

I like both, actually :smile: But for the first version I'd probably go with `--store-reports` or similar.

Comment (apahim) [inline avocado/plugins/diff.py]: I will use `--create-reports` then.

File: avocado/plugins/diff.py (hunk 1 above)
Comment (apahim) [inline avocado/plugins/diff.py]: ok

**Comment by apahim** [comment]
@ldoktor, updated with your requests.

File: avocado/plugins/diff.py [hunk 3]
```diff
@@ -0,0 +1,427 @@
...
+        parser = super(Diff, self).configure(parser)
+
+        parser.add_argument("jobids",
+                            default=[], nargs=2,
+                            metavar="<JOB>",
+                            help='Provide two Jobs, identified by their Job '
+                            '(partial) ID or test results directory.')
+
+        parser.add_argument('--html', type=str,
+                            dest='html_output',
```
Comment (ldoktor) [inline avocado/plugins/diff.py]: I wouldn't recommend using `dest` unless necessary (eg. when pairing two `store_true/store_false` arguments) as over time it becomes hard to follow. The option should always be equal to the variable....

File: avocado/plugins/diff.py [hunk 4]
```diff
@@ -0,0 +1,427 @@
...
+                            help='Generate and open a HTML report in your '
+                            'preferred browser. If no --html file is '
+                            'provided, we create a temporary file. '
+                            'Current: False.')
+
+        parser.add_argument('--diff-filter',
+                            dest='diff_filter',
+                            type=self._validate_filters,
+                            default=['all'],
+                            help='Diff filter: (no)cmdline,(no)time,'
```
Comment (ldoktor) [inline avocado/plugins/diff.py]: how about: comma separated filter of diff sections

File: avocado/plugins/diff.py [hunk 5]
```diff
@@ -0,0 +1,427 @@
...
+        if (('all' in args.diff_filter or 'cmdline' in args.diff_filter) and
+           'nocmdline' not in args.diff_filter):
+
+            cmdline1 = self._get_command_line(job1_dir)
+            cmdline2 = self._get_command_line(job2_dir)
+
+            if str(cmdline1) != str(cmdline2):
+                command_line_header = ['\n',
+                                       'COMMAND LINE\n',
+                                       '============\n']
```
Comment (ldoktor) [inline avocado/plugins/diff.py]: This would be controversal, but the convention in RST is to put 1-2 empty lines before and 1 line after the heading. For section (=) heading it's usually the 2 before and 1 after and I'd prefer that.

Also at least my vim colores the headings, making it more visible. As we don't want to make "coloring books", how about to use the `log.info` (blue) color?

I mean first sections are pretty easily to distinguish, but the more sections we have the harder it gets to find them...

File: avocado/plugins/diff.py [hunk 6]
```diff
@@ -0,0 +1,427 @@
...
+            if str(sysinfo_post1) != str(sysinfo_post2):
+                sysinfo_header_post = ['\n',
+                                       'SYSINFO POST\n',
+                                       '============\n']
+                job1_results.extend(sysinfo_header_post)
+                job1_results.extend(sysinfo_post1)
+                job2_results.extend(sysinfo_header_post)
+                job2_results.extend(sysinfo_post2)
+
+        if getattr(args, 'create_reports', False):
```
Comment (ldoktor) [inline avocado/plugins/diff.py]: well you know those options are present, you can use `args.create_reports` directly... (the same on other places) (can be tweaked after the release)

File: avocado/plugins/diff.py [hunk 7]
```diff
@@ -0,0 +1,427 @@
...
+                                                    for _ in job2_results),
+                                                    fromdesc=job1_id,
+                                                    todesc=job2_id)
+
+                with open(args.html_output, 'w') as f:
+                    f.writelines(job_diff_html.encode("utf-8"))
+
+                LOG.info('HTML diff file: %s' % args.html_output)
+
+            except Exception as e:
```
Comment (ldoktor) [inline avocado/plugins/diff.py]: the `e` is not really a nice name. Similar convention violations are all over this patch. Can you please run `pylint/pep8` and address them?

File: avocado/plugins/diff.py [hunk 8]
```diff
@@ -0,0 +1,427 @@
...
+                LOG.error(e)
+                sys.exit(exit_codes.AVOCADO_FAIL)
+
+        if getattr(args, 'open_browser', False):
+            setsid = getattr(os, 'setsid', None)
+            if not setsid:
+                setsid = getattr(os, 'setpgrp', None)
+            inout = file(os.devnull, "r+")
```
Comment (ldoktor) [inline avocado/plugins/diff.py]: opening a file using `file` is discouraged in python2.7 and forbidden in python3. Would you please use `open` insteald?

Comment (ldoktor) [inline avocado/plugins/diff.py]: also shouldn't we use `with open(...) as ...:` instead?

**Comment by ldoktor** [comment]
I'm done :smile: I haven't gone through the documentation (just skimmed through it), neither the selftests, but they can be improved later. The code looks quite good, there are few places for improvement, some little issues, but overall it looks almost mergeable.

File: avocado/plugins/diff.py (hunk 5 above)
Comment (apahim) [inline avocado/plugins/diff.py]: Well, I have my doubts we should follow rst here. The report in intended to be the most diff-able possible. Investing extra lines in nice looking headers makes them hard to show up in standard diffs, needing some manual tweaks. For now, I will revert this section identifiers and just add a (blue) color to the sections titles.

Comment (apahim) [inline avocado/plugins/diff.py]: Actually, I'm not going to use a color in it. Blue is used for diff snips. Yellow does no work well on white backgrounds. Also, I'm removing the `===` to not appear I'm trying to use rst format. It will be: `# SECTION`, no color.
